packages/docs/developer-resources/walkthroughs/web-dapp.md

```diff
@@ -0,0 +1,299 @@
+          </td>
+          <td>{proposal.vote ?? 'No vote yet'}</td>
+        </tr>
+      ))}
+    </tbody>
+  </table>
+)
+```
+
+### Locking Celo (optional)
+
+A prerequisite to [voting on Celo governance proposals](celo-owner-guide/voting-governance.md) is having locked Celo to vote with. We won't cover the various flows for locking, unlocking and relocking Celo in this tutorial but you can check the implementation in [Celo Tools](https://github.com/alexbharley/celo-tools) or take inspiration from the following script:
+
+```javascript
+const lockValue = new BigNumber(res.flags.value)
+
+const lockedGold = await this.kit.contracts.getLockedGold()
+const pendingWithdrawalsValue = await lockedGold.getPendingWithdrawalsTotalValue(address)
+const relockValue = BigNumber.minimum(pendingWithdrawalsValue, value)
+const lockValue = value.minus(relockValue)
+
+const txos = await lockedGold.relock(address, relockValue)
+for (const txo of txos) {
+  await kit.sendAndWaitForReceipt({ from: address })
+}
+```
+
+All you need to take care of in your React application is handling user input to select the amount to lock and handling errors in case the user tries to lock more CELO than they hold.
+
+It's also possible that users of your DApp already have locked CELO, so you might not need to worry about the complexity of permitting that operation.
+
+### Voting on a proposal
+
+To actually vote on a proposal we need to again interact with the [Governance.sol](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/governance/Governance.sol) smart contract. Our logic for handling a vote looks as follows:
+
+```typescript
+const vote = useCallback(
+  async (id: string, value: VoteValue) => {
+    // requireInitialised
+    const governance = await kit.contracts.getGovernance()
+    await (await governance.vote(id, value)).sendAndWaitForReceipt()
+    fetchProposals()
+  },
+  [kit, fetchProposals]
+)
+```
+
+How you handle calling that function is up to you. With [Celo Tools](https://github.com/alexbharley/celo-tools) we opted for simple upwards and downwards facing arrows to handle voting on proposals, however the data can be rendered however you'd prefer.
+
+Here's a simple example showing buttons for `Yes` or `No` votes when no vote has been cast.
+
+```javascript
+import { VoteValue } from '@celo/contractkit/lib/wrappers/Governance'
+
+return (
+  <tr>
+    <td>{proposal.id.toString()}</td>
+    <td>{proposal.passed ? 'Passed' : proposal.approved ? 'Approved' : 'Not approved'}</td>
+    <td>
+      <a href={proposal.descriptionURL} target="_blank">
+        Description link
+      </a>
+    </td>
+    <td>
+      {proposal.vote ? (
+        <span>{proposal.vote}</span>
+      ) : (
+        <div>
+          <button onClick={() => vote(proposal.id, VoteValue.Yes)}>Yes</button>
+          <button onClick={() => vote(proposal.id, VoteValue.No)}>No</button>
+        </div>
+      )}
+    </td>
+  </tr>
+)
+```
+
+## Best practices
```

Comment:
I love this section. We don't have to do this now, but I'd love to pull this out and have a Best Practices high-level section for reference outside of the tutorial.